core/src/test/scala/com/evolutiongaming/kafka/flow/ConsumerFlowSpec.scala

```diff
@@ -0,0 +1,178 @@
+package com.evolutiongaming.kafka.flow
+
+import cats.data.{NonEmptyMap, NonEmptySet}
+import cats.effect.{Resource, SyncIO}
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+import com.evolutiongaming.catshelper.LogOf
+import com.evolutiongaming.kafka.flow.kafka.Consumer
+import com.evolutiongaming.kafka.journal.ConsRecords
+import com.evolutiongaming.skafka.consumer.{ConsumerRecords, RebalanceListener}
+import com.evolutiongaming.skafka.{Offset, OffsetAndMetadata, Partition, Topic, TopicPartition}
+import munit.FunSuite
+
+import scala.concurrent.duration._
+
+class ConsumerFlowSpec extends FunSuite {
+
+  import ConsumerFlowSpec._
+
+  test("happy path") {
+    val topic = "test-topic"
+    val partition = Partition.unsafe(1)
+
+    val state = State(commands =
+      List(
+        Command.Assigned(NonEmptySet.of(TopicPartition(topic, partition))),
+        Command.Records(),
+        Command.Records()
+      )
+    )
+
+    val result = run(state, topic).unsafeRunSync()
+
+    assertEquals(
+      result.actions.reverse,
+      List(
+        Action.Subscribe(NonEmptySet.of(topic)),
+        Action.Add(NonEmptySet.of(partition -> offset)),
+        Action.Apply(),
+        Action.Apply()
+      )
+    )
+  }
+
+  test("multi topic subscription") {
+    val topic1 = "test-topic-1"
+    val topic2 = "test-topic-2"
+    val partition1 = Partition.unsafe(1)
+    val partition2 = Partition.unsafe(2)
+    val partition3 = Partition.unsafe(3)
+    val partition4 = Partition.unsafe(4)
+
+    val state = State(commands =
+      List(
+        Command.Assigned(
+          NonEmptySet.of(
+            TopicPartition(topic1, partition1),
+            TopicPartition(topic1, partition2),
+            TopicPartition(topic2, partition3),
+            TopicPartition(topic2, partition4)
+          )
+        ),
+        Command.Records(),
+        Command.Records()
+      )
+    )
+
+    val result = run(state, topic1, topic2).unsafeRunSync()
+
+    assertEquals(
+      result.actions.reverse,
+      List(
+        Action.Subscribe(NonEmptySet.of(topic1, topic2)),
+        Action.Add(NonEmptySet.of(partition1 -> offset, partition2 -> offset)),
+        Action.Add(NonEmptySet.of(partition3 -> offset, partition4 -> offset)),
+        Action.Apply(),
+        Action.Apply(),
+        Action.Apply(),
+        Action.Apply()
+      )
+    )
+  }
+
+}
+
+object ConsumerFlowSpec {
+
+  type F[A] = SyncIO[A]
```

Comment:
Could it be that you do not need `SyncIO` and `Ref` here as `ConsumerFlow` only depends on `BracketThrowable` and `LogOf`?

It looks to me that `StateT` would be more than enough and will actually make the test a bit more simple, but it may be just me :blush:

Reply:
Ah, okay. I mixed it up. You cannot easily replace `BracketThrowable` with `StateT` (you can only do it for `MonadThrowable`), but, as most of the logic is now in `RebalanceListener` does not require `BracketThrowable` and only requires `Monad`, it should be possible to cover it using `State` monad (as it does not throw any exceptions, `StateT` is not required).

Reply:
Done